File: src/indicators/sessions.py
"""src.indicators.sessions — Sessions High/Low math.

For each trading session (Asia / London / NY) on each day, find that session's
high and low, then:
  - a "ray" extends right from the high/low point until a later candle trades
    back to that level (a high is retested when a later high >= it; a low when a
    later low <= it), otherwise to the last bar;
  - "verticals" mark the session's start and end bar times.

Session windows are anchored to America/Chicago (the data's exchange tz),
DST-aware. This module is pure: OHLCV DataFrame in, levels out. Colors are a
frontend concern and deliberately live in the chart renderer, not here.

Returned dict (times are Unix seconds, UTC):
  {
    "sessions":  ["Asia", "London", "NY"],
    "rays":      [{"session","kind":"high"|"low","price","start","end"}, ...],
    "verticals": [{"session","kind":"start"|"end","time"}, ...],
  }
"""
from __future__ import annotations

import numpy as np
import pandas as pd

TZ = "America/Chicago"

# Session windows as minutes-from-midnight in TZ. A window "wraps" past midnight
# when end <= start (e.g. Asia 18:00 -> 03:00). Single source of truth.
SESSIONS = [
    {"name": "Asia",   "start": 18 * 60, "end": 3 * 60},
    {"name": "London", "start": 3 * 60,  "end": 8 * 60},
    {"name": "NY",     "start": 8 * 60,  "end": 17 * 60},
]

# Cap how many recent session instances we return, to bound output on
# timeframes that span many days.
MAX_SESSIONS = 60


def session_names() -> list[str]:
    return [s["name"] for s in SESSIONS]


def _empty() -> dict:
    return {"sessions": session_names(), "rays": [], "verticals": []}


def compute_sessions(df: pd.DataFrame, max_sessions: int = MAX_SESSIONS) -> dict:
    """OHLCV DataFrame (tz-aware UTC index) -> session H/L rays + verticals."""
    if df is None or df.empty:
        return _empty()

    local = df.index.tz_convert(TZ)
    minute = local.hour.values * 60 + local.minute.values
    highs = df["high"].to_numpy()
    lows = df["low"].to_numpy()
    times = df.index.view("int64") // 1_000_000_000  # Unix seconds (UTC)
    n = len(df)

    # Assign each bar to a session (windows here don't overlap -> first match).
    sess = np.full(n, -1, dtype=int)
    for si, s in enumerate(SESSIONS):
        wraps = s["end"] <= s["start"]
        if wraps:
            mask = (minute >= s["start"]) | (minute < s["end"])
        else:
            mask = (minute >= s["start"]) & (minute < s["end"])
        sess[(sess == -1) & mask] = si

    # Instance day: anchor a wrapping session's after-midnight bars to the day it
    # started (i.e. the previous calendar day).
    # Naive datetime64 (local wall-clock), one distinct value per local day.
    anchor = local.normalize().tz_localize(None).to_numpy().copy()
    for si, s in enumerate(SESSIONS):
        if s["end"] <= s["start"]:
            morning = (sess == si) & (minute < s["end"])
            anchor[morning] -= np.timedelta64(1, "D")

    keep = sess >= 0
    if not keep.any():
        return _empty()

    sub = pd.DataFrame({
        "pos": np.arange(n)[keep],
        "sess": sess[keep],
        "anchor": anchor[keep],
        "high": highs[keep],
        "low": lows[keep],
    })

    # One instance per (session, day): find start/end bar and where hi/lo formed.
    insts = []
    for (si, _day), g in sub.groupby(["sess", "anchor"], sort=False):
        hi_i = int(g["high"].values.argmax())
        lo_i = int(g["low"].values.argmin())
        insts.append({
            "sess": int(si),
            "start_pos": int(g["pos"].iloc[0]),
            "end_pos": int(g["pos"].iloc[-1]),
            "hi_price": float(g["high"].iloc[hi_i]),
            "hi_pos": int(g["pos"].iloc[hi_i]),
            "lo_price": float(g["low"].iloc[lo_i]),
            "lo_pos": int(g["pos"].iloc[lo_i]),
        })

    insts.sort(key=lambda x: x["start_pos"])
    insts = insts[-max_sessions:]

    def terminate(pos: int, price: float, up: bool) -> int:
        """Time of the first bar after `pos` that trades to `price`, else last."""
        if pos + 1 >= n:
            return int(times[-1])
        seg = highs[pos + 1:] if up else lows[pos + 1:]
        hit = seg >= price if up else seg <= price
        k = int(np.argmax(hit))
        return int(times[pos + 1 + k]) if hit[k] else int(times[-1])

    rays, verticals = [], []
    for it in insts:
        name = SESSIONS[it["sess"]]["name"]
        rays.append({
            "session": name, "kind": "high", "price": it["hi_price"],
            "start": int(times[it["hi_pos"]]),
            "end": terminate(it["hi_pos"], it["hi_price"], up=True),
        })
        rays.append({
            "session": name, "kind": "low", "price": it["lo_price"],
            "start": int(times[it["lo_pos"]]),
            "end": terminate(it["lo_pos"], it["lo_price"], up=False),
        })
        verticals.append({"session": name, "kind": "start", "time": int(times[it["start_pos"]])})
        verticals.append({"session": name, "kind": "end", "time": int(times[it["end_pos"]])})

    return {"sessions": session_names(), "rays": rays, "verticals": verticals}

File: chart/static/js/indicators/sessions.js
/* chart/static/js/indicators/sessions.js — Sessions High/Low indicator.
 *
 * For each session (Asia / London / NY) on each day:
 *   - draw a DASHED, color-coded horizontal ray from the session's high and low,
 *     extending right until a later candle trades back to the level, then stop;
 *   - draw DASHED vertical lines at the session's start and end, same color.
 *
 * Each session can be toggled independently (see `items` below + the control
 * panel). Session hours are anchored to America/Chicago (exchange tz), DST-aware.
 * Edit SESSIONS to change windows/colors — it's the single source of truth.
 */
(function () {
  // Session windows in Chicago local time, as minutes-from-midnight.
  // A window "wraps" past midnight when end <= start (e.g. Asia 18:00->03:00).
  const SESSIONS = [
    { name: 'Asia',   color: '#3f8ae0', start: 18 * 60, end: 3 * 60 },
    { name: 'London', color: '#e0a44e', start: 3 * 60,  end: 8 * 60 },
    { name: 'NY',     color: '#a06ee0', start: 8 * 60,  end: 17 * 60 },
  ];

  // Cap how many recent session instances we draw, to keep the series count
  // sane on timeframes that span many days.
  const MAX_SESSIONS = 60;

  // ---- Chicago-time helpers (DST-aware via Intl) ----------------------------
  const CHI = new Intl.DateTimeFormat('en-US', {
    timeZone: 'America/Chicago', hour12: false,
    year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', minute: '2-digit',
  });

  function chicagoParts(unixSec) {
    const o = {};
    for (const p of CHI.formatToParts(new Date(unixSec * 1000))) {
      if (p.type !== 'literal') o[p.type] = p.value;
    }
    return {
      year: +o.year, month: +o.month, day: +o.day,
      hour: (+o.hour) % 24, minute: +o.minute,
    };
  }

  function shiftDay(y, m, d, delta) {
    const dt = new Date(Date.UTC(y, m - 1, d));
    dt.setUTCDate(dt.getUTCDate() + delta);
    return `${dt.getUTCFullYear()}-${dt.getUTCMonth() + 1}-${dt.getUTCDate()}`;
  }

  function sessionFor(localMin) {
    for (const s of SESSIONS) {
      const wraps = s.end <= s.start;
      const inside = wraps
        ? (localMin >= s.start || localMin < s.end)
        : (localMin >= s.start && localMin < s.end);
      if (inside) return s;
    }
    return null;
  }

  // ---- Session computation --------------------------------------------------
  /* Group candles into session instances (one per session per day), tracking
   * each instance's start/end bar and where its high/low occurred. */
  function buildInstances(candles) {
    const byKey = new Map();
    for (let i = 0; i < candles.length; i++) {
      const c = candles[i];
      const p = chicagoParts(c.time);
      const localMin = p.hour * 60 + p.minute;
      const s = sessionFor(localMin);
      if (!s) continue;

      const wraps = s.end <= s.start;
      const dateKey = (wraps && localMin < s.end)
        ? shiftDay(p.year, p.month, p.day, -1)
        : `${p.year}-${p.month}-${p.day}`;
      const key = `${s.name}|${dateKey}`;

      let inst = byKey.get(key);
      if (!inst) {
        inst = {
          name: s.name, color: s.color,
          startIdx: i, endIdx: i,
          hi: -Infinity, hiIdx: i, lo: Infinity, loIdx: i,
        };
        byKey.set(key, inst);
      }
      inst.endIdx = i;
      if (c.high > inst.hi) { inst.hi = c.high; inst.hiIdx = i; }
      if (c.low < inst.lo) { inst.lo = c.low; inst.loIdx = i; }
    }
    return byKey;
  }

  /* First bar after `fromIdx` that trades to `price` (dir +1 = a later high
   * >= price; dir -1 = a later low <= price). Returns that bar's time, or the
   * last bar's time if never tested. */
  function terminate(candles, fromIdx, price, dir) {
    for (let j = fromIdx + 1; j < candles.length; j++) {
      if (dir > 0 ? candles[j].high >= price : candles[j].low <= price) {
        return candles[j].time;
      }
    }
    return candles[candles.length - 1].time;
  }

  /* -> { rays: [{name,color,price,start,end}], verticals: [{name,color,time}] } */
  function computeSessions(candles) {
    const insts = [...buildInstances(candles).values()]
      .sort((a, b) => a.startIdx - b.startIdx)
      .slice(-MAX_SESSIONS);

    const rays = [];
    const verticals = [];
    for (const inst of insts) {
      rays.push({
        name: inst.name, color: inst.color, price: inst.hi,
        start: candles[inst.hiIdx].time,
        end: terminate(candles, inst.hiIdx, inst.hi, +1),
      });
      rays.push({
        name: inst.name, color: inst.color, price: inst.lo,
        start: candles[inst.loIdx].time,
        end: terminate(candles, inst.loIdx, inst.lo, -1),
      });
      verticals.push({ name: inst.name, color: inst.color, time: candles[inst.startIdx].time });
      verticals.push({ name: inst.name, color: inst.color, time: candles[inst.endIdx].time });
    }
    return { rays, verticals };
  }

  // ---- Vertical-line canvas primitive --------------------------------------
  // Lightweight Charts has no native vertical line, so we draw them with a
  // series primitive that paints dashed verticals at given times/colors.
  class VLinesRenderer {
    constructor(src) { this._src = src; }
    draw(target) {
      const chart = this._src._chart;
      if (!chart) return;
      const ts = chart.timeScale();
      target.useMediaCoordinateSpace((scope) => {
        const ctx = scope.context;
        for (const ln of this._src._lines) {
          const x = ts.timeToCoordinate(ln.time);
          if (x === null) continue;
          const px = Math.round(x) + 0.5;
          ctx.save();
          ctx.strokeStyle = ln.color;
          ctx.lineWidth = 1;
          ctx.setLineDash([4, 4]);
          ctx.beginPath();
          ctx.moveTo(px, 0);
          ctx.lineTo(px, scope.mediaSize.height);
          ctx.stroke();
          ctx.restore();
        }
      });
    }
  }
  class VLinesPaneView {
    constructor(src) { this._renderer = new VLinesRenderer(src); }
    zOrder() { return 'bottom'; }
    renderer() { return this._renderer; }
  }
  class VLinesPrimitive {
    constructor() {
      this._lines = [];
      this._chart = null;
      this._requestUpdate = null;
      this._views = [new VLinesPaneView(this)];
    }
    attached(p) { this._chart = p.chart; this._requestUpdate = p.requestUpdate; }
    detached() { this._chart = null; this._requestUpdate = null; }
    updateAllViews() {}
    paneViews() { return this._views; }
    setLines(lines) {
      this._lines = lines;
      if (this._requestUpdate) this._requestUpdate();
    }
  }

  // ---- Indicator definition -------------------------------------------------
  window.IndicatorRegistry.register({
    id: 'sessions',
    label: 'Sessions H/L',
    description: 'Asia / London / NY session highs & lows with session boundaries',
    enabledByDefault: true,
    items: SESSIONS.map((s) => ({ id: s.name, label: s.name, color: s.color })),

    create(ctx) {
      const chart = ctx.chart;
      const series = [];
      const visible = { Asia: true, London: true, NY: true };
      let lastData = null;
      let lastTf = null;

      const vlines = new VLinesPrimitive();
      ctx.candleSeries.attachPrimitive(vlines);

      const clearSeries = () => {
        for (const s of series) chart.removeSeries(s);
        series.length = 0;
      };

      function draw() {
        clearSeries();
        const candles = lastData && lastData.candles;
        if (!candles || !candles.length || lastTf === '1d') {
          vlines.setLines([]);
          return;
        }
        const { rays, verticals } = computeSessions(candles);
        for (const r of rays) {
          if (!visible[r.name] || r.end <= r.start) continue;
          const s = chart.addLineSeries({
            color: r.color,
            lineWidth: 1,
            lineStyle: LightweightCharts.LineStyle.Dashed,
            lastValueVisible: false,
            priceLineVisible: false,
            crosshairMarkerVisible: false,
          });
          s.setData([
            { time: r.start, value: r.price },
            { time: r.end, value: r.price },
          ]);
          series.push(s);
        }
        vlines.setLines(
          verticals.filter((v) => visible[v.name]).map((v) => ({ time: v.time, color: v.color }))
        );
      }

      return {
        update(data, tf) { lastData = data; lastTf = tf; draw(); },
        setItemVisible(name, vis) { visible[name] = vis; draw(); },
        destroy() { clearSeries(); ctx.candleSeries.detachPrimitive(vlines); },
      };
    },
  });
})();
